src/parser.rs

```diff
@@ -108,12 +109,23 @@ fn parse_tag(iter: &mut Iter<Element>,
         Identifier(ref x) if options.blocks.contains_key(x) => {
             let end_tag = Identifier("end".to_owned() + &x);
             let mut children = vec![];
-            loop {
-                children.push(match iter.next() {
-                    Some(&Tag(ref tokens, _)) if tokens[0] == end_tag => break,
-                    None => break,
-                    Some(t) => t.clone(),
-                })
+            let mut nesting_depth = 0;
+            for t in iter {
+                if let &Tag(ref tokens, _) = t {
+                    match tokens[0] {
+                        ref n if n == tag => {
+                            nesting_depth += 1;
+                        },
+                        ref n if n == &end_tag && nesting_depth > 0 => {
+                            nesting_depth -= 1;
+                        },
+                        ref n if n == &end_tag && nesting_depth == 0 => {
+                            break
+                        },
+                        _ => {}
+                    }
+                };
```

Comment:
can you quickly elaborate (and maybe write a short comment) why you need to collect the nesting_depth here?